### PR TITLE
Migrate "ui" test suite from legacy compiletest-style directives `//` to `ui_test`-style directives `//@`

### PR DESCRIPTION
**Pull request by jieyouxu**
## Preface

There's an on-going effort to rewrite parts of or the entirety of compiletest
(<https://github.com/rust-lang/compiler-team/issues/536>). A step towards this involve migrating
ui tests to use the [`ui_test`](https://github.com/oli-obk/ui_test) framework, which involves
changing compiletest directives in `// <directive-name>` style to `ui_test` `//@ <directive-name>`
style (https://github.com/rust-lang/compiler-team/issues/512).

This PR aims to implement the directive-style change from `//` to `//@` for ui tests only and
make compiletest only accept `//@` directives in the "ui" test suite (only).

## Key Changes

1. All ui test `//` directives are replaced by `//@` directives.
2. Only accept `//@` directives for "ui" test suite.
3. Errors if a comment could be interpreted as a legacy-style `//` directive.

## Diff Generation 

The diff is generated by:

- Collecting directives from ui tests via hacking on compiletest.
- Using a migration tool to replace `//` directives in ui tests with `//@`.

### Reproduction Steps

0. Delete the temporary directory `$RUSTC_REPO_PATH/build/x86_64-apple-darwin/test/ui/__directive_lines` and the temporary file `$RUSTC_REPO_PATH/build/x86_64-apple-darwin/test/ui/__directive_lines.txt` (if you ran the collection script before).
1. Use the <https://github.com/jieyouxu/rust/tree/collect-test-directives> collect-test-directives
   script, which outputs temporary files recording headers occuring in each ui test.
   - You need to checkout this branch: `git checkout collect-test-directives`.
   - You might need to rebase on lastest master and ensure there are no conflicts.
   - You likely need to run `./x test tests/ui --stage 1 --force-rerun` to generate the temporary
     files consistently.
2. Checkout the `migrate-ui-test-directives` branch.
4. Run the migration tool <https://github.com/jieyouxu/compiletest-ui_test-header-migration>.
    - You will need to first generate a `migration_config.toml` via `cargo run -- generate-config` under `$CWD`.
    - Then, update `manual_directives = ["// should-fail"]` in `migration_config.toml`. This is required because the collection script doesn't deal with some special meta ui tests and there are no other `// should-fail` occurrences.
5. Check that the migration at least does not cause UI test failures if you change compiletest to
   accept `//@` directives for ui tests only.
   - `RUSTC_TEST_FAIL_FAST=1 ./x test tests/ui --stage 1 --bless`
6. Confirm that there is no difference after running the migration tool when you are on the
   `migrate-ui-test-directives` branch.

## Next Steps

- [x] ~~Need to implement some kind of warning or tidy script to help contributors catch old-style
  `// <directive-name>` directives, while only accepting `ui_test`-style `//@ <directive-name>`
  directives.~~ An error is emitted if a comment that could be interpreted as legacy-style test directive is encountered.
- [ ] Need to properly document this change in e.g. rustc-dev-guide (https://github.com/rust-lang/rustc-dev-guide/pull/1885).
    - [x] Add a `README.md` to `tests/ui` describing the directive style change.

### CODE REVIEW

**Comment by fmease** [comment]
Lol, did rustbot just give up?
r? compiler

**Comment by oli-obk** [comment]
cc @asquared31415 @pietroalbini

**Comment by asquared31415** [comment]
I no longer have the energy to work on my efforts for that, but I have pushed the rest of my changes to [my fork](https://github.com/asquared31415/rust/tree/compiletest_allow_uitest) and this was part of an effort to allow compiletest to properly parse ui_test style headers, not just change the syntax.  Feel free to use any of this code for any purpose.

I think it's a better move to just move to ui_test style comments without changing the actual commands.  However I would like to see a tidy warning and fix suggestion (that can auto apply on bless), to make sure that developers don't accidentally use the old style comments and the test doesn't actually test everything it's meant to.

**Comment by jieyouxu** [comment]
> I think it's a better move to just move to ui_test style comments without changing the actual commands. However I would like to see a tidy warning and fix suggestion (that can auto apply on bless), to make sure that developers don't accidentally use the old style comments and the test doesn't actually test everything it's meant to.

I agree that a tidy warning would be a good idea. I'm not entirely sure on the auto-apply fix suggestion though, because there might be unfortunate cases like

```rs
// This comment serves to mention that
// ignore-test is a very useful directive
// which ignores the test.
```

and then the test can become unintentionally ignored. (Even though is this the current behavior.)

**Comment by oli-obk** [comment]
It may be simpler to parse twice, once with `@` and once without, and error if any successfully parse without.

**Comment by jieyouxu** [comment]
> It may be simpler to parse twice, once with `@` and once without, and error if any successfully parse without.

I added a *very* dumb check, trying to parse a `//` comment as a legacy directive. Currently it simply bails and panics on the first such encountered directive.

File: tests/ui/abi/stack-protector.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
-// run-pass
-// only-x86_64-unknown-linux-gnu
-// revisions: ssp no-ssp
+//@ run-pass
+//@ only-x86_64-unknown-linux-gnu
+//@ revisions: ssp no-ssp
 // [ssp] compile-flags: -Z stack-protector=all
 // compile-flags: -C opt-level=2
```
Comment (oli-obk) [inline tests/ui/abi/stack-protector.rs:5]: manually verified that the latest version of the tool correctly handles this

**Comment by jieyouxu** [comment]
Sorry there's something weird with my diff, might have accidentally done something...

File: tests/ui/asm/inline-syntax.rs
[the remark this answers is not in the document]
Comment (jieyouxu) [reply tests/ui/asm/inline-syntax.rs:1]: > Pull this commit and the first commit into a separate PR maybe? Makes this PR cleaner

I opened #121045 for this, I'll drop the commits from this PR.

**Comment by oli-obk** [comment]
lgtm! Please squash.

will the collect-test-directives branch still work on top of this or does it include the relevant parts to skip tidy?

**Comment by jieyouxu** [comment]
> lgtm! Please squash.
> 
> will the collect-test-directives branch still work on top of this or does it include the relevant parts to skip tidy?

I'm going to squash these changes into two commits:

1. All the infrastructure part for accepting `//@` directives and warning/erroring on `//` directives.
2. Regenerate the auto-migration diff just to check if anything was missed or broken when considering the lastest changes to `master`.

**Comment by jieyouxu** [comment]
> lgtm! Please squash.
> 
> will the collect-test-directives branch still work on top of this or does it include the relevant parts to skip tidy?

I rebased the collect-test-directives branch on top of master, it was a clean rebase so all changed ui tests should be updated and no one touched compiletest (lol).

Using the rebased collect-test-directives branch, I will rerun it (i.e. the collection script) to collect any potentially new headers.

I will then regenerate the migration diff in a separate commit, so you can try to reproduce it on your end.

**Comment by oli-obk** [comment]
@bors p=512

I confirmed the tool does what it says it does and I reviewed the tool, so

@bors delegate+

**Comment by jieyouxu** [comment]
@oli-obk I fixed the headers parsing tests.

**Comment by jieyouxu** [comment]
Reblessed the migrated tests.

**Comment by oli-obk** [comment]
2024-02-16T14:37:30.5359776Z ##[group]Testing stage2 compiletest suite=coverage mode=coverage-map (x86_64-unknown-linux-gnu)
2024-02-16T14:37:30.6421168Z error: /checkout/tests/coverage/long_and_wide.rs:2: invalid line: ignore-tidy-linelength
2024-02-16T14:37:30.6667238Z 
2024-02-16T14:37:30.6667759Z thread 'main' panicked at src/tools/compiletest/src/lib.rs:578:9:
2024-02-16T14:37:30.6668363Z there are errors in tests
2024-02-16T14:37:30.6668934Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-02-16T14:37:30.6676805Z Build completed unsuccessfully in 0

**Comment by jieyouxu** [comment]
I forgot to conditionally not account for `ignore-tidy` and `ignore-tidy-*`,fixing now.

**Comment by oli-obk** [comment]
2024-02-16T15:21:20.2871561Z error: detected legacy-style directives in ui test: /checkout/tests/ui-fulldeps/session-diagnostic/invalid-variable.rs:2, please use `ui_test`-style directives `//@` instead:Some(
2024-02-16T15:21:20.2872632Z     (
2024-02-16T15:21:20.2872868Z         None,
2024-02-16T15:21:20.2873254Z         "compile-flags: --test",
2024-02-16T15:21:20.2873608Z     ),
2024-02-16T15:21:20.2873811Z )

**Comment by jieyouxu** [comment]
Missed a check for `mode == Mode::Ui` and `suite = "ui"`, fixing.

**Comment by jieyouxu** [comment]
Failure was due to not-specific enough checks for `ui` test suite (`ui-fulldeps` are also `Mode::Ui` tests but a different test suite). Should be fixed now.

**Comment by jieyouxu** [comment]
Probably missing a more specific condition for tidy check for in tidy, I'll double check.

**Comment by jieyouxu** [comment]
I fixed the incorrect `ignore-tidy` and `ignore-tidy-*` handling, and manually edited the `.fixed` of an ignored test.

**Comment by jieyouxu** [comment]
@oli-obk I am assuming that this **will** conflict with the current PR being bors'd once it merges since that rollup touches UI tests, we will have to redo the collect + migrate steps again.

**Comment by oli-obk** [comment]
you can use `@bors r=oli-obk` whenever you push an update, no need to wait on me

**Comment by oli-obk** [comment]
@bors r+

let's leave it in the queue in case the rollup fails

**Pull request merged**